Some minor UTF-8 spring-cleaning

dra27 opened the pull request:
While rebasing #13494, which adds `typo.utf8` to `.github/workflows/build-msvc.yml`, I noticed that #13526 had instead used `typo.non-ascii` for `.github/workflows/build-cross.yml` (there were already other files using non-ascii). The utf8 attribute is simply younger (it was added in #1828.

This pre-spring-cleaning rationalises things by requiring all text files to be valid UTF-8 (i.e. prohibit any old 8-bit encodings). Files which actually contain utf8 sequences still have to be marked as such.

dra27 committed:
Use C.UTF-8 for the test locale in check-typo

dra27 committed:
Require all text files to be valid UTF-8

Files which Git regards as binary are already skipped, but require _all_
non-binary files therefore to be valid UTF-8. The files still require
the utf8 attribute for UTF-8 to be _expected_.

dra27 committed:
Stop using non-ascii attribute

All the files using the non-ascii attribute are in fact UTF-8.

gasche commented:
If the intent is to say: "this file may content characters outside the ascii range", then I find `non-ascii` clearer than `utf8`. (I would guess that `utf8` means "this file is encoded in utf8", and I understand that this is now mandatory instead of an opt-in check.) Have you considered using only `non-ascii`, instead of using only `utf8`?

dra27 commented:
I hadn’t considered it, no - we could therefore remove the `utf8` attribute completely?

gasche commented:
That would be my preference, but maybe this nicety isn't worth the extra worth on your end. I'll approve and you'll decide.

dra27 committed:
Add check-typo to 5.x release branches!

dra27 committed:
Test check-typo PRs properly

Run check-typo over the entire tree if either .gitattributes or
check-typo itself have been altered.

dra27 committed:
Use non-ascii name rather than utf8

Rationale: all text files must be valid UTF-8 (and any UTF-8 is valid in
copyright headers, as before), but source files which use non-ASCII
characters (still forming valid UTF-8 sequences) must indicate
non-ascii.

dra27 commented:
Given that this all started from a confusion over the attribute, I thought it was worth persevering with over a coffee... of course, then I realised that the the previous state of the PR was actually going to break (cf. [9d07a2f](https://github.com/ocaml/ocaml/pull/13724/commits/9d07a2fa7c8210d4fa089b74bd7a54c343e1b588), so disappeared down a CI-improving rabbit whole 🐇

gasche reviewed (approved):
I'm approving, and will let @dra27 merge at his own pace -- I guess you are waiting for CI results and what not.
